source/decryptor/decryptor.c

```diff
@@ -292,16 +294,58 @@ static u8* FindNandCtr()
     return NULL;
 }
 
-u32 DumpPartition(char* filename, u32 offset, u32 size, u32 keyslot) {
+u32 SeekMagicNumber(u8* magic, u32 magiclen, u32 offset, u32 size, u32 keyslot)
+{
     DecryptBufferInfo info;
     u8* buffer = BUFFER_ADDRESS;
-    u8* ctrStart = FindNandCtr();
+    u32 found = (u32) -1;
+
+    if (ctrStart == NULL) {
+        ctrStart = FindNandCtr();
+        if (ctrStart == NULL)
+            return 1;
+    }
 
-    Debug("Dumping System NAND Partition. Size (MB): %u", size / (1024 * 1024));
+    info.keyslot = keyslot;
+    info.setKeyY = 0;
+    info.size = NAND_SECTOR_SIZE;
+    info.buffer = buffer;
+    for (u32 i = 0; i < 16; i++) {
+        info.CTR[i] = *(ctrStart + (0xF - i)); // The CTR is stored backwards in memory.
+    }
+    add_ctr(info.CTR, offset / 0x10);
+
+    u32 n_sectors = size / NAND_SECTOR_SIZE;
+    u32 start_sector = offset / NAND_SECTOR_SIZE;
+    for (u32 i = 0; i < n_sectors; i++) {
+        ShowProgress(i, n_sectors);
+        sdmmc_nand_readsectors(start_sector + i, 1, buffer);
+        DecryptBuffer(&info);
+        if(memcmp(buffer, magic, magiclen) == 0) {
+            found = (start_sector + i) * NAND_SECTOR_SIZE;
+            break;
+        }
+    }
+
+    ShowProgress(0, 0);
+    FileClose();
+
+    return found;
+}
+
+u32 DumpPartition(char* filename, u32 offset, u32 size, u32 keyslot)
+{
```

Comment:
If this function can be used to dump more than just partitions, can you rename it to reflect that?